WISE-Community/WISE
# Switch Angular $localize for in code translations

To test:
- Load site and make sure in-code translations are working.
- Test both develop and production builds.
- Make production builds of the 5 languages that have begun to be translated (es, pt, tr, zh-Hans, and zh-Hant) and check that translated strings appear.

Notes:
For new translations in .ts files going forward, we'll need to manually add an entry to `messages-code.xlf` for each new `$localize` instance. This is a slightly complicated process as of now, unfortunately. It involves first building one of the alternate language builds (e.g. `npm run build-dev-zh-Hans`) and noting the missing translation's id (which is automatically generated), then adding new `<trans-unit>` entry to `messages-code.xlf`. The process is outlined here: https://blog.ninja-squad.com/2019/12/10/angular-localize/ in the "i18n with $localize in TypeScript code" section. This is an example entry that you would add to `messages-code.xlf` after locating the id:
```
<trans-unit id="6480943972743237078">
  <source>Hello world.</source>
</trans-unit>
```
$localize message extraction will hopefully be added to the cli in angular 10 and then this will all be automated: https://github.com/angular/angular/pull/32912.

To run the `extract-i18n` script (used in pre-commit hook) which extracts i18n messages and merges `messages-code.xlf` into `messages.xlf`, you'll need to install these xml libraries:
```
sudo apt install xml-twig-tools
sudo apt install libxml2-utils
```

Closes #2450.
